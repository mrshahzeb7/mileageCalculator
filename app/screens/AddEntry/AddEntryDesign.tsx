import React, {useRef, useState} from 'react';
import {StyleSheet, View} from 'react-native';
import {icons} from '../../assets';
import {
  ComponentTree,
  DatePicker,
  DatePickerType,
  HeaderBackNavigation,
  Icon,
  Picker,
  RowInputProps,
  TextField,
} from '../../components';
import {colors, iconSizes, spacing} from '../../theme';
import {
  getCommaSeparatedValue,
  getDayForTimeLine,
  inputLabels,
  placeHolders,
  responsiveWidth,
} from '../../utils';
import {MainActions} from '../../redux';
import {useDispatch} from 'react-redux';
import {TimeLineSectionItem} from '../../components/TimeLineList/timeLineListContent';

const gasTypeList = [
  'Regular ',
  'Regular Two',
  'Regular Three',
  'Regular Four',
  'Regular Five',
  'Regular Six',
  'Regular Seven',
  'Regular Eight',
];
interface Props {
  closeModal: () => void;
}

const OdoMeterRightAccessory = () => {
  return <Icon size={iconSizes.medium} icon={icons.DropDownArrow} />;
};

export enum AddEntryFieldsKey {
  OdoMeter = 'OdoMeter',
  Gas = 'Gas',
  GasType = 'GasType',
  Price = 'Price',
  Cost = 'Cost',
  Date = 'Date',
  Time = 'Time',
}

export const AddEntryDesign: React.FC<Props> = ({closeModal}) => {
  const dispatch = useDispatch();
  const details = useRef<any>({
    [AddEntryFieldsKey.OdoMeter]: '',
    [AddEntryFieldsKey.Gas]: '',
    [AddEntryFieldsKey.GasType]: '',
    [AddEntryFieldsKey.Price]: '',
    [AddEntryFieldsKey.Cost]: '',
    [AddEntryFieldsKey.Date]: '',
    [AddEntryFieldsKey.Time]: '',
  });

  const [resetValues, _setResetValues] = useState<any>(false);

  const onOdoMeterChange = (text: string) =>
    updateValue(AddEntryFieldsKey.OdoMeter, text);

  const onGasChange = (text: string) =>
    updateValue(AddEntryFieldsKey.Gas, text);

  const onGasTypeChange = (text: string) =>
    updateValue(AddEntryFieldsKey.GasType, text);

  const onPricePerLiterChange = (text: string) =>
    updateValue(AddEntryFieldsKey.Price, text);

  const onCostChange = (text: string) =>
    updateValue(AddEntryFieldsKey.Cost, text);
  const onDateChange = (text: string) =>
    updateValue(AddEntryFieldsKey.Date, text);
  const onTimeChange = (text: string) =>
    updateValue(AddEntryFieldsKey.Time, text);

  const [componentList] = useState<RowInputProps[]>([
    {
      icon: icons.Meter,
      ChildOne: (
        <TextField
          keyboardType="number-pad"
          onChangeText={onOdoMeterChange}
          containerStyle={styles.odoMeter}
          placeholder={placeHolders.meter}
          helper="Last value: 66783 mi"
          RightAccessory={OdoMeterRightAccessory}
        />
      ),
    },
    {
      icon: icons.Gas,
      ChildOne: (
        <TextField
          keyboardType="number-pad"
          onChangeText={onGasChange}
          containerStyle={styles.fields}
          placeholder={placeHolders.gas}
        />
      ),
      ChildTwo: (
        <Picker
          list={gasTypeList}
          onSelect={onGasTypeChange}
          style={styles.fields}
          label={inputLabels.gasType}
        />
      ),
    },
    {
      icon: icons.Cost,
      ChildOne: (
        <TextField
          keyboardType="number-pad"
          onChangeText={onPricePerLiterChange}
          containerStyle={styles.fields}
          placeholder={placeHolders.price}
        />
      ),
      ChildTwo: (
        <TextField
          keyboardType="number-pad"
          onChangeText={onCostChange}
          containerStyle={styles.fields}
          placeholder={placeHolders.cost}
        />
      ),
    },
    {
      icon: icons.Calendar,
      ChildOne: (
        <DatePicker
          onSelect={onDateChange}
          style={styles.fields}
          type={DatePickerType.Date}
        />
      ),
      ChildTwo: (
        <DatePicker
          onSelect={onTimeChange}
          style={styles.fields}
          type={DatePickerType.Time}
        />
      ),
    },
  ]);

  const updateValue = (key: AddEntryFieldsKey, value: string) => {
    const newData = {...details.current};
    newData[key] = value;
    details.current = newData;
  };

  const onSave = () => {
    let date = getDayForTimeLine();

    const payload: TimeLineSectionItem = {
      meterReading: getCommaSeparatedValue(
        details?.current?.[AddEntryFieldsKey.OdoMeter] + '',
      ),
      gas: getCommaSeparatedValue(
        details?.current?.[AddEntryFieldsKey.Gas] + '',
      ),
      gasType: details?.current?.[AddEntryFieldsKey.GasType],
      price: getCommaSeparatedValue(
        details?.current?.[AddEntryFieldsKey.Price] + '',
      ),
      cost: getCommaSeparatedValue(
        details?.current?.[AddEntryFieldsKey.Cost] + '',
      ),
      date,
      time: details?.current?.[AddEntryFieldsKey.Time],
    };

    dispatch(MainActions.addEntryRequest(payload));
    closeModal?.();
  };

  const RightAction = (
    <Icon
      testID="HeaderSaveButton"
      onPress={onSave}
      color={colors.light}
      size={iconSizes.small}
      icon={icons.True}
    />
  );

  return (
    <View style={styles.container}>
      <HeaderBackNavigation
        title={'Refueling'}
        onLeftPress={closeModal}
        RightAction={RightAction}
      />
      <ComponentTree
        style={styles.componentTrees}
        key={resetValues}
        list={componentList}
      />
    </View>
  );
};

// styles
const styles = StyleSheet.create({
  container: {
    flex: 1,
    width: responsiveWidth(100),
    backgroundColor: colors.primary,
  },
  odoMeter: {
    flex: 1,
  },
  fields: {
    width: '41%',
  },
  componentTrees: {
    paddingTop: spacing.vertical.medium,
  },
});
