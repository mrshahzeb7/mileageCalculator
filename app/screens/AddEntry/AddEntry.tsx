import React from 'react';
import {AddButton} from '../../components';
import {Modal, useModalRef} from '../../components/Modal/Modal';
import {AddEntryDesign} from './AddEntryDesign';

interface Props {}

export const AddEntry: React.FC<Props> = () => {
  const modalRef = useModalRef();
  const onOpenModal = () => {
    modalRef?.current?.setVisibility?.(true);
  };

  const closeModal = () => {
    modalRef?.current?.setVisibility?.(false);
  };

  return (
    <>
      <AddButton onPress={onOpenModal} />
      <Modal ref={modalRef}>
        <AddEntryDesign closeModal={closeModal} />
      </Modal>
    </>
  );
};
