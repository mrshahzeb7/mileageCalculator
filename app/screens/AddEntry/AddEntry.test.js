import 'react-native';
import React from 'react';
import {AddEntry} from './AddEntry';
import {AddEntryDesign} from './AddEntryDesign';
import renderer from 'react-test-renderer';
import {
  SafeAreaProvider,
  initialWindowMetrics,
} from 'react-native-safe-area-context';

const Parent = ({children}) => (
  <SafeAreaProvider initialMetrics={initialWindowMetrics}>
    {children}
  </SafeAreaProvider>
);

it('renders AddEntry correctly', () => {
  const tree = renderer
    .create(
      <Parent>
        <AddEntry />
      </Parent>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders AddEntryDesign correctly', () => {
  const tree = renderer
    .create(
      <Parent>
        <AddEntryDesign />
      </Parent>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
