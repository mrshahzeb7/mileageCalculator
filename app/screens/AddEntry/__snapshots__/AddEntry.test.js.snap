// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders AddEntry correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    {
      "flex": 1,
    }
  }
/>
`;

exports[`renders AddEntryDesign correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    {
      "flex": 1,
    }
  }
/>
`;
